### PR TITLE
CastPlayer bug fixes for issues/4168

### PR DESCRIPTION
**Pull request by mkaflowski**
Couple bug fixes for issue:
[https://github.com/google/ExoPlayer/issues/4168](https://github.com/google/ExoPlayer/issues/4168)
Mostly wrong currentWindowIndex.

### CODE REVIEW

File: extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java [hunk 1]
```diff
@@ -145,42 +145,57 @@ public CastPlayer(CastContext castContext) {
     pendingSeekWindowIndex = C.INDEX_UNSET;
     pendingSeekPositionMs = C.TIME_UNSET;
     updateInternalState();
+
+    if (remoteMediaClient != null) {
+      try {
```
Comment (mkaflowski) [inline extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java]: 1. Wrong getCurrentWindowIndex after setting media sources when CastPlayer is connected to Chromecast device.

File: extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java [hunk 2]
```diff
@@ -145,42 +145,57 @@ public CastPlayer(CastContext castContext) {
     pendingSeekWindowIndex = C.INDEX_UNSET;
     pendingSeekPositionMs = C.TIME_UNSET;
     updateInternalState();
+
```
Comment (mkaflowski) [inline extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java]: Fix for:
1. Wrong getCurrentWindowIndex after setting media sources when CastPlayer is connected to Chromecast device.

File: extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java [hunk 3]
```diff
@@ -145,42 +145,57 @@ public CastPlayer(CastContext castContext) {
...
+   * @param item       The item to load.
    * @param positionMs The position at which the playback should start in milliseconds relative to
-   *     the start of the item at {@code startIndex}. If {@link C#TIME_UNSET} is passed, playback
-   *     starts at position 0.
+   *                   the start of the item at {@code startIndex}. If {@link C#TIME_UNSET} is passed, playback
+   *                   starts at position 0.
    * @return The Cast {@code PendingResult}, or null if no session is available.
    */
   public PendingResult<MediaChannelResult> loadItem(MediaQueueItem item, long positionMs) {
-    return loadItems(new MediaQueueItem[] {item}, 0, positionMs, REPEAT_MODE_OFF);
```
Comment (mkaflowski) [inline extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java]: Fix for:
3. ArrayIndexOutOfBoundsException: length=2; index=2 when setting new media sources and some other problems with wron current position.

File: extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java [hunk 4]
```diff
@@ -145,42 +145,57 @@ public CastPlayer(CastContext castContext) {
...
+   *                   starts at position 0.
    * @param repeatMode The repeat mode for the created media queue.
    * @return The Cast {@code PendingResult}, or null if no session is available.
    */
   public PendingResult<MediaChannelResult> loadItems(MediaQueueItem[] items, int startIndex,
-      long positionMs, @RepeatMode int repeatMode) {
+                                                     long positionMs, @RepeatMode int repeatMode) {
     if (remoteMediaClient != null) {
       positionMs = positionMs != C.TIME_UNSET ? positionMs : 0;
       waitingForInitialTimeline = true;
```
Comment (mkaflowski) [inline extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java]: Fix for:
3. ArrayIndexOutOfBoundsException: length=2; index=2 when setting new media sources and some other problems with wron current position.

File: extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java [hunk 5]
```diff
@@ -734,12 +753,12 @@ private static int fetchRepeatMode(RemoteMediaClient remoteMediaClient) {
 
   /**
    * Retrieves the current item index from {@code mediaStatus} and maps it into a window index. If
-   * there is no media session, returns 0.
```
Comment (mkaflowski) [inline extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java]: Fix for:
2. I get 0 from getCurrentWindowIndex() which is not true.

File: extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java [hunk 6]
```diff
@@ -472,28 +489,29 @@ public int getCurrentWindowIndex() {
...
-        : currentTimeline.getPreviousWindowIndex(getCurrentWindowIndex(), repeatMode, false);
+            : currentTimeline.getPreviousWindowIndex(getCurrentWindowIndex(), repeatMode, false);
   }
 
   // TODO: Fill the cast timeline information with ProgressListener's duration updates.
   // See [Internal: b/65152553].
   @Override
   public long getDuration() {
     return currentTimeline.isEmpty() ? C.TIME_UNSET
-        : currentTimeline.getWindow(getCurrentWindowIndex(), window).getDurationMs();
```
Comment (mkaflowski) [inline extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java]: Update for fix for:
2. I get 0 from getCurrentWindowIndex() which is not true.

File: extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java [hunk 7]
```diff
@@ -508,20 +526,20 @@ public int getBufferedPercentage() {
...
-        ? 0
-        : duration == 0 ? 100 : Util.constrainValue((int) ((position * 100) / duration), 0, 100);
+            ? 0
+            : duration == 0 ? 100 : Util.constrainValue((int) ((position * 100) / duration), 0, 100);
   }
 
   @Override
   public boolean isCurrentWindowDynamic() {
     return !currentTimeline.isEmpty()
-        && currentTimeline.getWindow(getCurrentWindowIndex(), window).isDynamic;
```
Comment (mkaflowski) [inline extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java]: Update for fix for:
2. I get 0 from getCurrentWindowIndex() which is not true.

File: extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java [hunk 8]
```diff
@@ -508,20 +526,20 @@ public int getBufferedPercentage() {
...
+            && currentTimeline.getWindow(getCurrentWindowIndex() < 0 ? 0 : getCurrentWindowIndex(), window).isDynamic;
   }
 
   @Override
   public boolean isCurrentWindowSeekable() {
     return !currentTimeline.isEmpty()
-        && currentTimeline.getWindow(getCurrentWindowIndex(), window).isSeekable;
```
Comment (mkaflowski) [inline extensions/cast/src/main/java/com/google/android/exoplayer2/ext/cast/CastPlayer.java]: Update for fix for:
2. I get 0 from getCurrentWindowIndex() which is not true.

**Comment by tonihei** [comment]
We usually merge pull requests into our dev-v2 branch. Can you recommit the change into this branch? Thanks!

**Pull request closed without merging**